### PR TITLE
feat: transliteration function

### PR DESCRIPTION
**Pull request by RomainMuller**
When a `.tgz` package is uploaded into an S3 bucket, the `jsii-rosetta
transpile` function is automatically run on the assembly contained
therein.

When sucessful, the transliterated assemblies are automatically uploaded
to the same S3 bucket, with a different object key.

The initial implementation only performs Python generation (this is
hard-coded and will be changed later).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: .projenrc.js [hunk 1]
```diff
@@ -92,8 +98,14 @@ const project = new AwsCdkConstructLibrary({
 
   // run tests from .js -- otherwise lambda bundlers get confused
   testdir: 'src/__tests__',
+
+  // Exclude handler images from TypeScript compier path
+  excludeTypescript: ['resources/**'],
 });
 
+// Required while we vendor-in jsii-rosetta to a pre-release version
+project.addFields({ resolutions: { '@jsii/spec': './vendor/jsii-spec.tgz' } });
```
Comment (eladb) [inline .projenrc.js:107]: I'd add the vendor deps here so they are co located:

```js
project.addDevDeps('jsii-rosetta@./vendor/jsii-rosetta.tgz');
```

File: src/back-end/index.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export * from './transliterator';
```
Comment (eladb) [inline src/back-end/index.ts:1]: Isn't it all backend?

Also, why two words?

File: src/back-end/transliterator/handler.lambda.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      Key: record.s3.object.key,
+      VersionId: record.s3.object.versionId,
+    }).promise();
+
+    const workdir = await fs.mkdtemp(path.join(os.tmpdir(), 'workdir'));
+    try {
+      const tarball = path.join(workdir, `${packageName.replace('@', '').replace('/', '-')}-${packageVersion}.tgz`);
+      await fs.writeFile(tarball, object.Body!);
+      await new Promise<void>((ok, ko) => {
+        const npmInstall = spawn('npm', ['install', '--ignore-scripts', '--no-bin-links', '--no-save', tarball], {
```
Comment (eladb) [inline src/back-end/transliterator/handler.lambda.ts:54]: Add a note that ignore scripts is required for security reasons

File: src/back-end/transliterator/handler.lambda.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        });
+      });
+      const packageDir = path.join(workdir, 'node_modules', packageName);
+
+      await transliterateAssembly(
+        [packageDir],
+        [TargetLanguage.PYTHON], // TODO: allow configuring this
+      );
+
+      const key = record.s3.object.key.replace(/\/[^/]+$/, '/assembly-python.json');
```
Comment (eladb) [inline src/back-end/transliterator/handler.lambda.ts:78]: Add a comment showing how the final key will look like

File: src/back-end/transliterator/index.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+   * @default RetentionDays.TEN_YEARS
+   */
+  readonly logRetention?: RetentionDays;
+}
+
+export class Transliterator extends Construct {
+  public constructor(scope: Construct, id: string, props: TransliteratorProps) {
+    super(scope, id);
+
+    const lambda = new Handler(this, 'Default', {
```
Comment (eladb) [inline src/back-end/transliterator/index.ts:23]: I try to call these per their class name (`Handler`) so the code will be future compatible with c8 🤪

**Review by eladb — COMMENTED** [review]
Looks really good. LGTM. I'll let @NetaNir and others add their approval.

File: src/back-end/index.ts (hunk 2 above)
Comment (RomainMuller) [reply src/back-end/index.ts:1]: The app will be modelling both back-end and front-end resources (i.e: the WebApp & CloudFront distro).

File: src/back-end/transliterator/handler.lambda.ts (hunk 3 above)
Comment (RomainMuller) [reply src/back-end/transliterator/handler.lambda.ts]: It is actually mentioned in the test (that was not present when you initially reviewed this).

File: src/back-end/transliterator/index.ts (hunk 5 above)
Comment (RomainMuller) [reply src/back-end/transliterator/index.ts:23]: You mean for the construct ID? Is this not the typical case where you _want_ the `Handler` instance to be the `defaultChild`?

**Review by MrArnoldPalmer — APPROVED** [review]
looks really good. Pretty easy to reason about imo. Question are we planning to add more languages to this handler or more handlers for each language using the same trigger?

**Comment by RomainMuller** [comment]
> Question are we planning to add more languages to this handler or more handlers for each language using the same trigger?

At this stage I wanna experiment with it a bit before I commit to a decision. I reckon the "most reliable" way would be 1 handler per language (based off the same code, though), but most cost-effective likely will be one for all languages.

**Pull request merged**